Fix the pydantic logging validator

maxdebayser opened the pull request:
Prevent the validator from logging unused extra fields before object is validated by the pydantic validator.

During the routing based on the request content where pydantic tries several classes for deserialization, this would result in spurious logs.

For example, during in a request like this one:

```
curl http://localhost:8000/v1/embeddings \
  -H "Content-Type: application/json" \
  -d '{
    "input": "The rampant curiosity for what a new Tool album would sound like has gotten more and more exaggerated with each release cycle.",
    "model": "sentence-transformers/all-MiniLM-L12-v2"
  }'
```

vLLM would log
```
WARNING 01-24 19:52:15 protocol.py:66] The following fields were present in the request but ignored: {'input'}
```

This is because pydantic tries to deserialize the JSON either as an `EmbeddingCompletionRequest` or an `EmbeddingChatRequest`. During the validation of the `EmbeddingChatRequest` it shows this log because the field only exists in `EmbeddingCompletionRequest` . This could lead users to believe that the `input` field was ignored while processing the request.

cc: @DarkLight1337 , @njhill (since you touched this code recently)

maxdebayser committed:
Fix the pydantic logging validator

Prevent the validator from logging unused extra fields
before object is validated by the pydantic validator.

During the routing based on the request content where
pydantic tries several classes for deserialization, this
would result in spurious logs.

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

DarkLight1337 reviewed (approved):
Thanks for fixing! Also cc @mgoin

DarkLight1337 commented:
Hmm, looks like this is causing a bunch of tests to fail. For example:

```
[2025-01-24T20:53:45Z] Traceback (most recent call last):
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/uvicorn/protocols/http/httptools_impl.py", line 409, in run_asgi
[2025-01-24T20:53:45Z]     result = await app(  # type: ignore[func-returns-value]
[2025-01-24T20:53:45Z]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
[2025-01-24T20:53:45Z]     return await self.app(scope, receive, send)
[2025-01-24T20:53:45Z]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/fastapi/applications.py", line 1054, in __call__
[2025-01-24T20:53:45Z]     await super().__call__(scope, receive, send)
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/applications.py", line 112, in __call__
[2025-01-24T20:53:45Z]     await self.middleware_stack(scope, receive, send)
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/middleware/errors.py", line 187, in __call__
[2025-01-24T20:53:45Z]     raise exc
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/middleware/errors.py", line 165, in __call__
[2025-01-24T20:53:45Z]     await self.app(scope, receive, _send)
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/middleware/cors.py", line 85, in __call__
[2025-01-24T20:53:45Z]     await self.app(scope, receive, send)
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/middleware/exceptions.py", line 62, in __call__
[2025-01-24T20:53:45Z]     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/_exception_handler.py", line 53, in wrapped_app
[2025-01-24T20:53:45Z]     raise exc
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/_exception_handler.py", line 42, in wrapped_app
[2025-01-24T20:53:45Z]     await app(scope, receive, sender)
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/routing.py", line 715, in __call__
[2025-01-24T20:53:45Z]     await self.middleware_stack(scope, receive, send)
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/routing.py", line 735, in app
[2025-01-24T20:53:45Z]     await route.handle(scope, receive, send)
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/routing.py", line 288, in handle
[2025-01-24T20:53:45Z]     await self.app(scope, receive, send)
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/routing.py", line 76, in app
[2025-01-24T20:53:45Z]     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/_exception_handler.py", line 53, in wrapped_app
[2025-01-24T20:53:45Z]     raise exc
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/_exception_handler.py", line 42, in wrapped_app
[2025-01-24T20:53:45Z]     await app(scope, receive, sender)
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/starlette/routing.py", line 73, in app
[2025-01-24T20:53:45Z]     response = await f(request)
[2025-01-24T20:53:45Z]                ^^^^^^^^^^^^^^^^
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/fastapi/routing.py", line 301, in app
[2025-01-24T20:53:45Z]     raw_response = await run_endpoint_function(
[2025-01-24T20:53:45Z]                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/fastapi/routing.py", line 212, in run_endpoint_function
[2025-01-24T20:53:45Z]     return await dependant.call(**values)
[2025-01-24T20:53:45Z]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/openai/api_server.py", line 364, in show_available_models
[2025-01-24T20:53:45Z]     models_ = await handler.show_available_models()
[2025-01-24T20:53:45Z]               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/openai/serving_models.py", line 115, in show_available_models
[2025-01-24T20:53:45Z]     ModelCard(id=base_model.name,
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/pydantic/main.py", line 212, in __init__
[2025-01-24T20:53:45Z]     validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
[2025-01-24T20:53:45Z]                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/openai/protocol.py", line 51, in __log_extra_fields__
[2025-01-24T20:53:45Z]     result = handler(data)
[2025-01-24T20:53:45Z]              ^^^^^^^^^^^^^
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/openai/protocol.py", line 69, in __log_extra_fields__
[2025-01-24T20:53:45Z]     data.keys() - field_names)
[2025-01-24T20:53:45Z]     ^^^^^^^^^
[2025-01-24T20:53:45Z]   File "/usr/local/lib/python3.12/dist-packages/pydantic/main.py", line 856, in __getattr__
[2025-01-24T20:53:45Z]     raise AttributeError(f'{type(self).__name__!r} object has no attribute {item!r}')
[2025-01-24T20:53:45Z] AttributeError: 'ModelPermission' object has no attribute 'keys'
```

maxdebayser committed:
Don't validate non-dict objects

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
trigger ci

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Merge branch 'upstream_main' into fix_misleading_logs

maxdebayser committed:
trigger ci

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Merge branch 'upstream_main' into fix_misleading_logs

maxdebayser committed:
Fix BatchRequestInput pydantic validation

When not enough input fields are given, the
deserialization of the body is ambiguous.

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>
